torch_frame/data/multi_nested_tensor.py

```diff
@@ -78,85 +78,161 @@ def from_tensor_mat(
+                    torch.tensor(index, device=self.device),
+                    dim=0,
+                )
+                return col_index(multi_nested_tensor, index[1])
+            else:
+                raise NotImplementedError
+        # index is for row indexing.
+        elif isinstance(index, slice):
+            # tensor[i:j]
+            # Returns MultiNestedTensor
+            return self.slice(index, dim=0)
         elif isinstance(index, int):
-            # Support negative indices
-            if index < 0:
-                index = self.num_rows + index
-            if index < 0 or index >= self.num_rows:
-                raise IndexError("Index out of bounds!")
-            # Get i-th row. Returns MultiNestedTensor
-            start_idx = index * self.num_cols
-            end_idx = (index + 1) * self.num_cols + 1
-            offset = self.offset[start_idx:end_idx]
-            values = self.values[offset[0]:offset[-1]]
-            offset = offset - offset[0]
-            return MultiNestedTensor(num_rows=1, num_cols=self.num_cols,
-                                     values=values, offset=offset)
+            # tensor[i]
+            # Returns MultiNestedTensor
+            return self.single_index_select(index, dim=0)
         elif isinstance(index, Tensor) and index.ndim == 1:
+            # tensor[[i0, i1, i2, ..., iN]]
+            # Returns MultiNestedTensor
             return self.index_select(index, dim=0)
         elif isinstance(index, List):
+            # tensor[[i0, i1, i2, ..., iN]]
+            # Returns MultiNestedTensor
             return self.index_select(
-                torch.tensor(index, device=self.values.device), dim=0)
+                torch.tensor(index, device=self.device),
+                dim=0,
+            )
         else:
-            raise RuntimeError("Advanced indexing not supported yet.")
+            raise NotImplementedError("Advanced indexing not supported yet.")
 
-    def __len__(self):
-        return self.num_rows
+    def slice(self, slice: slice, dim: int) -> 'MultiNestedTensor':
+        if dim == 0 or dim == -3:
+            return self.row_slice(slice)
+        elif dim == 1 or dim == -2:
+            return self.col_slice(slice)
 
     def index_select(self, index: Tensor, dim: int) -> Tensor:
         if dim == 0:
             return self._row_index_select(index)
         else:
-            raise RuntimeError(
-                f"index_select for dim={dim} not supported yet.")
-
-    def _row_index_select(self, index: Tensor) -> Tensor:
-        index = index.clone()
-        # Support negative indices
-        neg_idx = index < 0
-        index[neg_idx] = self.num_rows + index[neg_idx]
-        if index.min() < 0 or index.max() >= self.num_rows:
-            raise IndexError("Index out of bounds!")
+            raise RuntimeError(f"Unsupported dim={dim} for index_select.")
+
+    def row_slice(self, slice: slice) -> 'MultiNestedTensor':
```

Comment:
this is known as narrow. Let's try to follow PyTorch terminology.

Reply:
I checked `narrow` (https://pytorch.org/docs/stable/generated/torch.narrow.html) but the arguments are different from `slice`.
It needs to pass `start` and `length`. There is no `step` and `length` is different from `end`.

Reply:
That's true, but you can compute end from start and length. Overall, I just think it is nicer when we inherit PyTorch functions here so that it feels like a normal tensor.

Reply:
I see. But it cannot model the step. Shall we add narrow by calling row_slice internally?

Reply:
Sounds good.